fix(acp): preserve file attachment metadata during session replay

liorshk opened the pull request:
## Problem

Fixes #9217 https://github.com/anomalyco/opencode/issues/9217

When reloading an ACP session, file attachments were not being replayed correctly:
- `file` type parts in stored messages were not being sent to ACP clients at all
- Images lost their original filenames (always showed as "image")
- Binary files (PDFs, etc.) were not replayed
- Text files appeared as inline text instead of as file attachments
- Synthetic text parts (injected for LLM context) were incorrectly sent to ACP clients

## Solution

Updated `processMessage()` in `agent.ts` to properly handle `file` parts during session replay:

- **`file://` URLs** → Replayed as `resource_link` blocks
- **Images** → Replayed as `image` blocks with preserved original filename
- **Text files** (`text/*`, `application/json`) → Replayed as `resource` blocks with decoded text
- **Binary files** → Replayed as `resource` blocks with blob data

Also updated the `prompt()` method to:
- Preserve original filenames from `resource_link.name` and `image.uri`
- Handle binary `resource` blocks (PDFs, etc.) by storing them as file parts with data URLs
- Check for truthy `resource.text` instead of just property existence

## Changes

- Add `file` part handling in `processMessage()` for session replay
- Filter synthetic text parts from ACP output (`!part.synthetic`)
- Preserve filename from `resource_link.name` when converting to internal parts
- Handle binary `resource.blob` blocks in `prompt()`

rekram1-node commented on packages/opencode/src/session/message-v2.ts:
```diff
@@ -438,14 +438,37 @@ export namespace MessageV2 {
-              mediaType: part.mime,
-              filename: part.filename,
-            })
+          if (part.type === "file") {
```
im skeptical of the updates here, what are the reasons here?

This file shouldnt really need updates to fix acp stuff

liorshk replied:
You're right - the message-v2.ts changes were unnecessary for this fix.
The initial approach tried to handle binary content detection at the LLM message conversion layer, but that was the wrong place to solve this problem. The actual issue is purely in the ACP layer: when replaying a session, OpenCode's file parts weren't being converted back to proper ACP content blocks (image, resource, resource_link).
The current fix keeps all changes in agent.ts.

I updated the PR and the PR description.

liorshk committed:
fix(acp): preserve file attachment metadata during session replay

This PR fixes the issue where file attachments were not being correctly replayed during ACP session reload.

Changes to processMessage():
- Add handling for 'file' parts to correctly replay attachments
- file:// URLs -> resource_link blocks
- Images -> image blocks with preserved filename
- Text files -> resource blocks with decoded text
- Binary files -> resource blocks with blob data
- Filter synthetic text parts from ACP output

Changes to prompt():
- Preserve filenames from resource_link.name and image.uri
- Handle binary resource blocks (PDFs, etc.) as file parts
- Check for truthy resource.text instead of just property existence

rekram1-node commented:
/review

liorshk committed:
Addressing style comments

liorshk committed:
Merge branch 'dev' into fix/acp-file-attachments-clean

rekram1-node committed:
Change condition from 'synthetic' to 'ignored'
